### ISSUE TITLE
clang-tidy returned non-zero status 1 cause action to fail

### ISSUE THREAD

**Issue by shenxianpeng** [shenxianpeng/cpp-linter-action#24]
Running the following command has output but its return code is 1 cause the action to fail.

```bash
subprocess.CalledProcessError: Command '['clang-tidy-11', '-checks=boost-*,bugprone-*,performance-*,readability-*,portability-*,modernize-*,clang-analyzer-*,cppcoreguidelines-*', '--export-fixes=clang_tidy_output.yml', 'demo.cpp']' returned non-zero exit status 1.
```
More details please refer to this action https://github.com/shenxianpeng/test-cpp-linter-action/runs/3782914731

I did a manual test, it looks good, it return 0, so I was confused why it failed. any suggestions?

```bash
gitpod /workspace/test-cpp-linter-action $ clang-tidy-11 -checks=boost-*,bugprone-*,performance-*,readability-*,portability-*,modernize-*,clang-analyzer-*,cppcoreguidelines-* --export-fixes=clang_tidy_output.yml demo.cpp
957 warnings generated.
/workspace/test-cpp-linter-action/demo.cpp:3:10: warning: inclusion of deprecated C++ header 'stdio.h'; consider using 'cstdio' instead [modernize-deprecated-headers]
#include <stdio.h>
         ^~~~~~~~~
         <cstdio>
/workspace/test-cpp-linter-action/demo.cpp:8:5: warning: use a trailing return type for this function [modernize-use-trailing-return-type]
int main(){
~~~ ^
auto       -> int
/workspace/test-cpp-linter-action/demo.cpp:10:13: warning: statement should be inside braces [readability-braces-around-statements]
    for (;;) break;
            ^
             {
/workspace/test-cpp-linter-action/demo.cpp:13:5: warning: do not call c-style vararg functions [cppcoreguidelines-pro-type-vararg]
    printf("Hello world!\n");
    ^
Suppressed 953 warnings (949 in non-user code, 4 with check filters).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
gitpod /workspace/test-cpp-linter-action $ echo $?
0
```

**Comment by 2bndy5** [comment]
Well the `checks=True` keyword arg to `subprocess.run()` is what causes any non-zero exit code to be raised as an exception (like you are seeing). Clang-tidy is probably exiting with a non-zero code to intentionally draw attention from the user (I need to investigate further).

Honestly, using the `checks=True` was only meant to satisfy the code-inspector suggestions. So, we can probably just take that out.

**Comment by 2bndy5** [comment]
Looking at the log a bit closer, I see this
> Error while trying to load a compilation database:
Could not auto-detect compilation database for file "demo.cpp"
No compilation database found in /github/workspace or any parent directory
fixed-compilation-database: Error while opening fixed database: No such file or directory
json-compilation-database: Error while opening JSON database: No such file or directory
Running without flags.

This might have caused clang-tidy to exit with non-zero code. I did change the compile_commands.json into [compile_flags.txt](https://github.com/shenxianpeng/cpp-linter-action/blob/master-py/demo/compile_flags.txt)  because it seemed that clang-tidy would only look for the compile_commands.json in a certain location (CMake generated this file in the specified _build_ directory). This might be related to the fact that you're not using the git checkout action beforehand.

**Comment by 2bndy5** [comment]
Another line from the log:
> Found compiler error(s).

This is definitely why clang-tidy exits with a non-zero code. It very likely related to not using git checkout action beforehand.

I suppose we could add `-extra-arg="-Wall -Werror"` to the `cmds` list  when this action has to manually download the file in question. But this would then be applied to all files that were manually downloaded.

I have to run some local tests...

**Comment by 2bndy5** [comment]
TLDR: We should mandate users do a git checkout before running this action.

I can reproduce this locally if I delete or rename the _demo.hpp_ file (which instigates the `Found compiler error(s)` message). I have run into this message in the past when the _demo.hpp_ had malformed syntax in it. I should note that _demo.cpp_ does say to `#include "demo.hpp"`. However, since you didn't do the git checkout action beforehand, the _demo.hpp_ file wouldn't exist, thus a compilation failure.

**Comment by 2bndy5** [comment]
I ran the changes locally while the file _demo.hpp_ was not present. It generated the expected usual output with the addition of this new part
<details open>
<summary><strong>demo\demo.cpp:2:10:</strong> error: [clang-diagnostic-error]

> 'demo.hpp' file not found
</summary><p>

```cpp
#include "demo.hpp"
         ^
```
</p>
</details>

**Comment by shenxianpeng** [comment]
After adding `- uses: actions/checkout@v2` to checkout repo, it works. 👍

**Comment by shenxianpeng** [comment]
Test passed (related test [commit](https://github.com/shenxianpeng/test-cpp-linter-action/commit/6a389d6bfa11b48cd19af520de9b4bed67752372)) and closed.
